### ISSUE TITLE
[CI] UpdateHealthInfoCacheIT testMasterFailure failing

### ISSUE THREAD

**Issue by Tim-Brooks** [elastic/elasticsearch#97213]
This has failed a few times in the last six months with the same failure message:

```
Expected: <DiskHealthInfo[healthStatus=GREEN, cause=null]>
     but: was null
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:956)
		at org.junit.Assert.assertThat(Assert.java:923)
		at org.elasticsearch.health.UpdateHealthInfoCacheIT.assertResultsCanBeFetched(UpdateHealthInfoCacheIT.java:153)
		at org.elasticsearch.health.UpdateHealthInfoCacheIT.lambda$testHealthNodeFailOver$4(UpdateHealthInfoCacheIT.java:88)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1136)
		... 38 more
	Suppressed: java.lang.AssertionError: 
Expected: <DiskHealthInfo[healthStatus=GREEN, cause=null]>
     but: was null
		at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
		at org.junit.Assert.assertThat(Assert.java:956)
		at org.junit.Assert.assertThat(Assert.java:923)
		at org.elasticsearch.health.UpdateHealthInfoCacheIT.assertResultsCanBeFetched(UpdateHealthInfoCacheIT.java:153)
		at org.elasticsearch.health.UpdateHealthInfoCacheIT.lambda$testHealthNodeFailOver$4(UpdateHealthInfoCacheIT.java:88)
		at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1136)
		... 38 more
	Suppressed: java.lang.AssertionError: 
Expected: <DiskHealthInfo[healthStatus=GREEN, cause=null]>
```

**Build scan:**
https://gradle-enterprise.elastic.co/s/atedejz57asri/tests/:server:internalClusterTest/org.elasticsearch.health.UpdateHealthInfoCacheIT/testMasterFailure

**Reproduction line:**
```
./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.health.UpdateHealthInfoCacheIT.testMasterFailure" -Dtests.seed=20B3752FEA453C10 -Dtests.locale=ro -Dtests.timezone=America/Antigua -Druntime.java=20
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.health.UpdateHealthInfoCacheIT&tests.test=testMasterFailure

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: <DiskHealthInfo[healthStatus=GREEN, cause=null]>
     but: was null

  at __randomizedtesting.SeedInfo.seed([20B3752FEA453C10:F859FCA181F45392]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.health.UpdateHealthInfoCacheIT.assertResultsCanBeFetched(UpdateHealthInfoCacheIT.java:131)
  at org.elasticsearch.health.UpdateHealthInfoCacheIT.lambda$testMasterFailure$5(UpdateHealthInfoCacheIT.java:103)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1172)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1145)
  at org.elasticsearch.health.UpdateHealthInfoCacheIT.testMasterFailure(UpdateHealthInfoCacheIT.java:103)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

**Comment by gmarouli** [comment]
Looking at this failure there are two things that pop to mind first:
- This could be a timing issue since the polling interval is 10 seconds and the `assertBusy` times out at 10 seconds too.
- There is a race somewhere and we lose the update.

After a first look, I am leaning towards the second. Observing the timeline of this test, we see:

- [2023-06-28T16:28:58,794] `node_t0` is elected as master.
- [2023-06-28T16:28:58,997] `node_t4` is selected as the health node.
- [2023-06-28T16:29:00,130] we introduce the disruption by restarting `node_t0`.
- [2023-06-28T16:29:00,409] `node_t0` is elected as master.
- [2023-06-28T16:29:00,789] `node_t3` is selected as the current health node.

The health node or the master node change should have triggered an immediate resend of the health info, we cancel a currently scheduled job and we start a new one... So, theoretically 10 seconds should have been more than enough to receive a cluster update and send the data.... 

I wi continue investigating the race hypothesis.

**Comment by gmarouli** [comment]
Furthermore, reading the issue a bit more carefully, I see that the first error mentioned is from a different test `testHealthNodeFailOver`. Because of this, I will focus on the general reset mechanism when a health node or master node get reset, since they share similar mechanics:

```
if (healthNodeChanged || masterNodeChanged) {
            // On health node or on master node changes, the health node might be reset so the reported
            // health info gets reset to null, to ensure it will be resent.
            lastSeenHealthNode.set(currentHealthNode == null ? null : currentHealthNode.getId());
            lastReportedDiskHealthInfo.set(null);
            if (logger.isDebugEnabled()) {
                String reason;
                if (healthNodeChanged && masterNodeChanged) {
                    reason = "the master node and the health node";
                } else if (healthNodeChanged) {
                    reason = "the health node";
                } else {
                    reason = "the master node";
                }
                logger.debug(
                    "Resetting the health monitoring because {} changed, current health node is {}.",
                    reason,
                    currentHealthNode == null ? null : format("[%s][%s]", currentHealthNode.getName(), currentHealthNode.getId())
                );
            }
        }
```

**Comment by gmarouli** [comment]
Enabled debug logging to get more data during the failures: https://github.com/elastic/elasticsearch/pull/97523

**Comment by gmarouli** [comment]
No failures yet...

**Comment by gmarouli** [comment]
No failures yet...

**Comment by gmarouli** [comment]
No failures yet

**Comment by gmarouli** [comment]
@andreidan I am passing this to you after your request, I am afraid that there is a race indeed and I believe the race is somewhere in this part of the code https://github.com/elastic/elasticsearch/blob/2076183deea2bad1af73eabe3e3213badecc1a9a/server/src/main/java/org/elasticsearch/health/node/LocalHealthMonitor.java#L174C5-L210C6. See https://github.com/elastic/elasticsearch/issues/97213#issuecomment-1623266162 & https://github.com/elastic/elasticsearch/issues/97213#issuecomment-1628592149

I have tried to reproduce it locally but I failed. I enabled debug logging so we can see exactly what happens but unfortunately there haven't been any failures on main in the last 3 weeks. I considered finding the issue through analysis but I wanted to have some more concrete evidence via a reproduction path.
